Update docs for 2.15 (#260)
canonical/mlflow-operator: docs/tutorial/mlflow-kubeflow.rst

+-----------+---------+
| Component | Version |
+-----------+---------+
|   MLflow  |   2.15  |
+-----------+---------+
|  Kubeflow |   1.9   |
+-----------+---------+

> Welcome to this tutorial on getting started with Charmed MLflow alongside Charmed Kubeflow. If you would like to deploy Kubeflow by itself, see our tutorial on `Charmed Kubeflow <https://charmed-kubeflow.io/docs/get-started-with-charmed-kubeflow>`_.

Reviewer: ```suggestion
This tutorial gets you started with Charmed MLflow alongside Charmed Kubeflow (CKF).
```
Add hyperlink to the CKF home page under the "Charmed Kubeflow (CKF)" text

This tutorial assumes you will be deploying Kubeflow and MLflow on a public cloud Virtual Machine (VM) with the following specs:

- Runs Ubuntu 20.04 (focal) or later.
- Has at least 4 cores, 32GB RAM and 200GB of disk space available.
- Is connected to the internet for downloading the required snaps and charms.

We'll also assume that you have a laptop that meets the following conditions:

- Has an SSH tunnel open to the VM with port forwarding and a SOCKS proxy. To see how to set this up, see `How to setup SSH VM Access <https://charmed-kubeflow.io/docs/how-tosetup-ssh-vm-access-with-port-forwarding>`_.

> - Runs Ubuntu 22.04 (focal) or later.

Author: You right 20.04 still works

This tutorial assumes you will be deploying Kubeflow and MLflow on a public cloud Virtual Machine (VM) with the following specs:

- Runs Ubuntu 20.04 (focal) or later.
- Has at least 4 cores, 32GB RAM and 200GB of disk space available.

> - Is connected to the internet for downloading the required snaps and charms.

Reviewer: I would remove this one. Internet connection should be always assumed, no?

> We'll also assume that you have a laptop that meets the following conditions:

Reviewer: ```suggestion
Also, your machine should meet the following requirements:
```

.. code-block:: bash

   juju deploy kubeflow --trust  --channel=2.15/stable

> Don't forget to set credentials for your Kubeflow deployment:

Reviewer: ```suggestion
Set credentials for your Kubeflow deployment:
```

.. code-block:: bash

   juju config dex-auth static-username=admin
   juju config dex-auth static-password=admin

Deploy Resource Dispatcher
--------------------------

> Next, let's deploy the resource dispatcher. The resource dispatcher is an optional component which will distribute Kubernetes objects related to MLflow credentials to all user namespaces in Kubeflow. This means that all your Kubeflow users can access the MLflow model registry from their namespaces. To deploy the dispatcher, run the following command:

Reviewer: ```suggestion
Next, deploy the resource dispatcher. The resource dispatcher is an optional component which distributes Kubernetes objects related to MLflow credentials to all user namespaces in Kubeflow. This means that all your Kubeflow users can access the MLflow model registry from their namespaces. To deploy the dispatcher, run the following command:
```

.. code-block:: bash

   juju deploy resource-dispatcher --channel 2.0/stable --trust

> This will deploy the latest stable version of the dispatcher. See `Resource Dispatcher on GitHub <https://github.com/canonical/resource-dispatcher>`_ for more info. Now we must relate the dispatcher to MLflow:

Reviewer: ```suggestion
This deploys the latest stable version of the dispatcher. See `Resource Dispatcher on GitHub <https://github.com/canonical/resource-dispatcher>`_ for more info. Now we must relate the dispatcher to MLflow:
```
Let's use present tense as much as possible

.. code-block:: bash

   juju integrate mlflow-server:secrets resource-dispatcher:secrets
   juju integrate mlflow-server:pod-defaults resource-dispatcher:pod-defaults

> Because we also want to deploy sorted MLflow models using KServe, in the next step we will create needed relations:

Reviewer: ```suggestion
To deploy sorted MLflow models using KServe, create the required relations as follows:
```

Don't be surprised if some of the components' statuses change to ``blocked`` or ``error`` every now and then. This is expected behaviour, and these statuses should resolve by themselves as the bundle configures itself. However, if components remain stuck in the same error states, consult the troubleshooting steps below.

Be patient, it can take up to half an hour for all those charms to download and initialise. In the meantime, why not try our `Juju tutorial <https://juju.is/docs/juju/get-started-with-juju>`_?

Integrate MLflow with Notebook
------------------------------

In this section, we're going to create a notebook server in Kubeflow and connect it to MLflow. This will allow our notebook logic to talk to MLflow in the background. Let's get started.

> First, to be able to use MLflow credentials in your Kubeflow notebook, visit the dashboard at ``http://10.64.140.43.nip.io/`` and fill the username and password which you configured in the previous section e.g. ``admin`` and ``admin``.

Reviewer 2: I think that we should mention that the IP could be different, depends on what IP range was configured for metallb and if user has already something else deployed. e.g. I deployed cos first and then kubeflow, so for me, it's http://10.64.140.44.nip.io/. WDYT?

Author: Added note to the doc

Don't be surprised if some of the components' statuses change to ``blocked`` or ``error`` every now and then. This is expected behaviour, and these statuses should resolve by themselves as the bundle configures itself. However, if components remain stuck in the same error states, consult the troubleshooting steps below.

> Be patient, it can take up to half an hour for all those charms to download and initialise. In the meantime, why not try our `Juju tutorial <https://juju.is/docs/juju/get-started-with-juju>`_?

Reviewer: ```suggestion
It can take up to half an hour for all charms to be downloaded and initialised. 
```
Add this info as a note box

Integrate MLflow with Kubeflow Dashboard
----------------------------------------

> You can integrate your charmed MLflow deployment with Kubeflow dashboard by running following commands: 

Reviewer: ```suggestion
You can integrate your charmed MLflow deployment with the Kubeflow dashboard by running following commands: 
```

.. code-block:: bash

   juju integrate mlflow-server:ingress istio-pilot:ingress
   juju integrate mlflow-server:dashboard-links kubeflow-dashboard:links

> Now you should see the MLflow tab in the left sidebar of your Kubeflow Dashboard at ``http://10.64.140.43.nip.io/``.

Reviewer: ```suggestion
Now you should see the MLflow tab in the left sidebar of your Kubeflow dashboard at ``http://10.64.140.43.nip.io/``.
```
We use lowercase before for "dashboard". Is it an own name? Should we use CAPS instead?

Reviewer: Can we add ``http://10.64.140.43.nip.io/`` as a code block so the user can copy and paste?

.. note:: 
   

>    The address of your Kubeflow Dashboard may differ depending on your setup. You can always check the IP by running: 

Reviewer: ```suggestion
   The address of your Kubeflow dashboard may differ depending on your setup. You can always check its URL by running: 
```
Again, upper or lowercase for dashboard? Also, isn't it an URL instead of IP?

   
   .. code-block:: bash
      
      microk8s kubectl -n kubeflow get svc istio-ingressgateway-workload -o jsonpath='{.status.loadBalancer.ingress[0].ip}'

Integrate MLflow with Notebook
------------------------------

In this section, we're going to create a notebook server in Kubeflow and connect it to MLflow. This will allow our notebook logic to talk to MLflow in the background. Let's get started.

> First, to be able to use MLflow credentials in your Kubeflow notebook, visit the dashboard at ``http://10.64.140.43.nip.io/`` and fill the username and password which you configured in the previous section e.g. ``admin`` and ``admin``.

Reviewer: Add a hyperlink to that section.

.. note:: 
   
   The address of your Kubeflow dashboard may differ depending on your setup. You can always check its URL by running: 
   
   .. code-block:: bash
      
      microk8s kubectl -n kubeflow get svc istio-ingressgateway-workload -o jsonpath='{.status.loadBalancer.ingress[0].ip}'

Integrate MLflow with Notebook
------------------------------

In this section, you are going to create a Kubeflow notebook server and connect it to MLflow. 

First, to be able to use MLflow credentials in your Kubeflow notebook, go to the MLflow dashboard at ``http://10.64.140.43.nip.io/`` 
and use the username and password you configured in the previous :ref:`kubeflow-section` section.
For example, ``admin`` and ``admin``.

Click on ``Start setup`` to setup the Kubeflow user for the first time.

Select ``Finish`` to finish the process.

Now a Kubernetes namespace is created for your user. 

Now go back to the dashboard. From the left panel, choose ``Ǹotebooks``. 
Select ``+New Notebook``.

At this point, name the notebook as you prefer, and choose the desired image and resource limits. 
For example, you can use the following details:

1. ``Name``: ``test-notebook``.

> 2. Expand the *Custom Notebook* section and for ``image``, select ``kubeflownotebookswg/jupyter-tensorflow-full:v1.7.0``.
> 

Reviewer 3: ```suggestion
2. Expand the *Custom Notebook* section and for ``image``, select ``kubeflownotebookswg/jupyter-tensorflow-full:v1.9.0``.

```

From the menu, choose the ``Clone a Repository`` option.

Now insert this repository address ``https://github.com/canonical/charmed-kubeflow-uats.git``.

> This will clone a whole ``charmed-kubeflow-uats`` repository onto the notebook server. The cloned repository will be a folder on the server, with the same name as the remote repository. Go inside the folder and after that, choose the ``tests/notebooks`` sub-folder.

Reviewer: ```suggestion
This clones a whole ``charmed-kubeflow-uats`` repository onto the notebook server. The cloned repository is a folder on the server, with the same name as the remote repository. Go inside the folder and after that, choose the ``tests/notebooks`` sub-folder.
```

> There you will find following folders:

Reviewer: ```suggestion
There you find following folders:
```

> - ``mlflow-kserve``: demonstrates how to talk to MLflow and KServe from inside a notebook. This example will train a simple ML model, store it in MLflow, deploy it with KServe from MLflow and run inference.

Reviewer: ```suggestion
- ``mlflow-kserve``: demonstrates how to talk to MLflow and KServe from inside a notebook. This example trains a simple ML model, stores it in MLflow, deploys it with KServe from MLflow and runs inference.
```
